source/simulator/assembly.cc

```diff
@@ -814,6 +817,13 @@ namespace aspect
       }
 
 
```

Comment:
It seems natural to me to update to matrix-free right-hand side right after assembling. Maybe the `evaluate_material_model()` can move, however. Where are the material model values set and reset for each nonlinear iteration? In `core.cc`? Maybe we don't need to do this every time we assemble.